### PR TITLE
Add pNFT support to Auction House

### PR DESCRIPTION
**Pull request by Flawm**
**Summary**
- Add support for pNFTs to `cancel`, `sell` and `execute_sale` logic
- New testing methods in testing-utils in order to create & mint pNFTs
- Add tests for pNFT functionality in Auction House

The way to build an instruction with pNFT support is to add in the remaining accounts to to the ix, in the case of execute_sale, they go after the creator accounts

### CODE REVIEW

**Review by blockiosaurus — APPROVED** [review]

File: auction-house/program/src/execute_sale/mod.rs [hunk 1]
```diff
@@ -34,6 +39,7 @@ pub struct ExecuteSale<'info> {
 
     /// CHECK: Validated in execute_sale_logic.
     /// Metaplex metadata account decorating SPL mint account.
+    //@TODO: re-enable this later #[account(mut)]
```
Comment (febo) [inline auction-house/program/src/execute_sale/mod.rs:42]: Is this meant to be commented out?

File: auction-house/program/src/execute_sale/mod.rs [hunk 2]
```diff
@@ -507,6 +551,7 @@ pub struct AuctioneerExecuteSale<'info> {
...
```
Comment (febo) [inline auction-house/program/src/execute_sale/mod.rs:554]: Is this meant to be commented out?

File: auction-house/program/src/sell/mod.rs [hunk 3]
```diff
@@ -23,6 +29,7 @@ pub struct Sell<'info> {
 
     /// CHECK: Verified through CPI
     /// Metaplex metadata account decorating SPL mint account.
+    //@TODO: remove #[account(mut)]
```
Comment (febo) [inline auction-house/program/src/sell/mod.rs:32]: Should this be commented out?

File: auction-house/program/src/sell/mod.rs [hunk 4]
```diff
@@ -140,6 +172,7 @@ pub struct AuctioneerSell<'info> {
 
     /// CHECK: Validated by assert_metadata_valid.
     /// Metaplex metadata account decorating SPL mint account.
+    //@TODO: remove #[account(mut)]
```
Comment (febo) [inline auction-house/program/src/sell/mod.rs:175]: Should this be commented out?

File: token-metadata/program/src/processor/metadata/transfer.rs [hunk 5]
```diff
@@ -110,8 +110,12 @@ fn transfer_v1(program_id: &Pubkey, ctx: Context<Transfer>, args: TransferArgs)
     if let Some(master_edition) = ctx.accounts.edition_info {
         assert_owned_by(master_edition, program_id)?;
     }
+
+    // if a blank rule set comes in we don't need to check this, catch all
     if let Some(authorization_rules) = ctx.accounts.authorization_rules_info {
-        assert_owned_by(authorization_rules, &mpl_token_auth_rules::ID)?;
+        if !authorization_rules.data_is_empty() {
```
Comment (febo) [inline token-metadata/program/src/processor/metadata/transfer.rs:116]: When can this happen?

**Review by febo — DISMISSED** [review]

**Review by febo — DISMISSED** [review]

File: auction-house/program/src/execute_sale/mod.rs (hunk 1 above)
Comment (blockiosaurus) [reply auction-house/program/src/execute_sale/mod.rs:42]: Yup, this is the breaking change we'll add back in later.

File: token-metadata/program/src/processor/metadata/transfer.rs (hunk 5 above)
Comment (Flawm) [reply token-metadata/program/src/processor/metadata/transfer.rs:116]: If you want to see an example, they're in the test cases, create a pNFT with asset data rule_set: Option<None>

**Review by blockiosaurus — APPROVED** [review]

**Review by blockiosaurus — APPROVED** [review]

**Review by blockiosaurus — APPROVED** [review]

**Review by febo — APPROVED** [review]

**Pull request merged**